pkg/virt-controller/watch/vsock_test.go

```diff
@@ -0,0 +1,121 @@
+package watch
+
+import (
+	"fmt"
+	"math"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/controller"
+	"kubevirt.io/kubevirt/tests"
+)
+
+var _ = Describe("VSOCK", func() {
+	var m *cidsMap
+
+	newRandomVMIsWithORWithoutVSOCK := func(totalVMINum, vsockVMINum int) []*virtv1.VirtualMachineInstance {
+		var vmis []*virtv1.VirtualMachineInstance
+		for i := 0; i < totalVMINum; i++ {
+			vmi := tests.NewRandomVMI()
+			if i < vsockVMINum {
+				cid := uint32(i + 3)
+				vmi.Status.VSOCKCID = &cid
+			}
+			vmis = append(vmis, vmi)
+		}
+		return vmis
+	}
+
+	BeforeEach(func() {
+		m = newCIDsMap()
+	})
+
+	DescribeTable("Syncing CIDs from exsisting VMIs",
+		func(totalVMINum, vsockVMINum int) {
+			vmis := newRandomVMIsWithORWithoutVSOCK(totalVMINum, vsockVMINum)
+			for _, vmi := range vmis {
+				fmt.Println(vmi.Status.VSOCKCID)
+			}
+			m.Sync(vmis)
+			Expect(m.cids).To(HaveLen(vsockVMINum))
+			Expect(m.reverse).To(HaveLen(vsockVMINum))
+			for _, vmi := range vmis {
+				if vmi.Status.VSOCKCID == nil {
+					Expect(m.cids).NotTo(HaveKey(controller.VirtualMachineInstanceKey(vmi)))
+				} else {
+					Expect(m.cids).To(HaveKey(controller.VirtualMachineInstanceKey(vmi)))
+					Expect(m.reverse).To(HaveKey(*vmi.Status.VSOCKCID))
+				}
+			}
+		},
+		Entry("When all VMIs have VSOCK enabled", 3, 3),
+		Entry("When some VMIs have VSOCK enabled", 5, 3),
+	)
+
+	DescribeTable("Allocating CIDs for VMIs",
+		func(num int) {
+			vmis := newRandomVMIsWithORWithoutVSOCK(num, 0)
+			for _, vmi := range vmis {
+				Expect(m.Allocate(vmi)).To(Succeed())
+				Expect(vmi.Status.VSOCKCID).NotTo(BeNil())
+				Expect(*vmi.Status.VSOCKCID).To(BeNumerically(">=", 3))
+			}
```

Comment:
Let's add here a check which ensures that we can't find an allocated cid twice. Like iterating over VMIs, putting the CIDs in a map and fail if the key is already in the map.

Reply:
Done